### PR TITLE
SALTO-820 convert custom record type instances to types

### PR DESCRIPTION
**Pull request by dantal4**
Convert custom record type instances to types. This is a preparation for fetching custom record instances.

---

_Additional context for reviewer_
This change effects several filters & validator, because until this point they were all assuming that we fetch/deploy only instances. 

---
_Release Notes_: 
Netsuite Adapter:
- Convert custom record type instances to types.

---
_User Notifications_: 
Netsuite: convert custom record type instances to types.
- The old `customrecordtype` instances will be deleted from the workspace
- References to custom record types will be replaced all over the workspace.
- The new types will be stored in `netsuite/CustomRecords` folder. Here is one for example:
```
type netsuite.customrecord1 {
  serviceid scriptid {
    _required = true
  }
  string internalId {
    _hidden_value = true
  }
  string custom_custrecord_text_field {
    scriptid = "custrecord_text_field"
    fieldtype = "TEXT"
    ...
  }
  netsuite.account custom_custrecord_ref_to_account {
    scriptid = "custrecord_ref_to_account"
    fieldtype = "SELECT"
    selectrecordtype = "-112"
    ...
  }
  scriptid = "customrecord1"
  ...
  permissions = {
    permission = {
      role1 = {
        permittedlevel = "VIEW"
        permittedrole = netsuite.role.instance.role1
        index = 0
      }
    }
  }
  metadataType = "customrecordtype"
}
```

### CODE REVIEW

File: packages/netsuite-adapter/src/elements_source_index/elements_source_index.ts [hunk 1]
```diff
@@ -41,31 +40,43 @@ const createIndexes = async (elementsSource: ReadOnlyElementsSource):
...
         : _.mapValues(serviceIdRecords, ({ elemID }) => ({ elemID, lastFetchTime }))
     )
   }
 
-  const updateInternalIdsIndex = (element: InstanceElement): void => {
-    const { internalId, isSubInstance } = element.value
+  const updateInternalIdsIndex = (element: InstanceElement | ObjectType): void => {
+    const { internalId, isSubInstance } = isInstanceElement(element)
+      ? element.value
+      : element.annotations
```
Comment (omrilit) [inline packages/netsuite-adapter/src/elements_source_index/elements_source_index.ts:66]: I wonder whether the following code should be extracted to a dedicated func, WDYT?
```
isInstanceElement(element)
  ? element.value
  : element.annotations
```

File: packages/netsuite-adapter/src/filters/internal_ids/sdf_internal_ids.ts [hunk 2]
```diff
@@ -167,10 +204,15 @@ const filterCreator: FilterCreator = ({ client }) => ({
...
-      delete element.value.internalId
+    const changesData = changes.map(getChangeData)
+    getSupportedInstances(changesData).forEach(inst => {
+      delete inst.value[INTERNAL_ID]
+    })
+    changesData.filter(isObjectType).filter(isCustomRecordType).forEach(type => {
+      delete type.annotations[INTERNAL_ID]
+    })
+    changesData.filter(isField).filter(field => isCustomRecordType(field.parent)).forEach(field => {
+      delete field.parent.annotations[INTERNAL_ID]
```
Comment (omrilit) [inline packages/netsuite-adapter/src/filters/internal_ids/sdf_internal_ids.ts:215]: nit: Please consider to do something in that way so you'll call only once per type and not multipled by the amount of its fields.
```suggestion
    uniqBy(changesData.filter(isField).map(field => field.parent), type => type.elemID.name).filter(type => isCustomRecordType(type)).forEach(customRecordType => {
      delete customRecordType.annotations[INTERNAL_ID]
```

File: packages/netsuite-adapter/src/service_url/instances_urls.ts [hunk 3]
```diff
@@ -35,24 +36,27 @@ const getScriptIdToInternalId = async (
   )
 }
 
-export const setInstancesUrls = async (
+export const setElementsUrls = async <T extends Element>(
```
Comment (omrilit) [inline packages/netsuite-adapter/src/service_url/instances_urls.ts:39]: please modify the file name

**Review by omrilit — COMMENTED** [review]
Wow this was a HUUUGE PR!
Couldn't find any logical issue but only nits about naming and tiny performance things.
As we agreed, this PR should be **tested & merged** with other related PRs to unify all of the noise to a single release.
GREAT wotk @dantal4 !

File: packages/netsuite-adapter/src/change_validators/instance_changes.ts [hunk 4]
```diff
@@ -13,15 +13,14 @@
...
-} from '@salto-io/adapter-api'
+import { isInstanceElement, ChangeValidator, getChangeData, isAdditionOrModificationChange } from '@salto-io/adapter-api'
+import { isCustomTypeElement } from '../types'
 
 const changeValidator: ChangeValidator = async changes => (
   changes
     .filter(isAdditionOrModificationChange)
     .map(getChangeData)
-    .filter(elem => !isInstanceElement(elem))
+    .filter(elem => !isInstanceElement(elem) && !isCustomTypeElement(elem))
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/change_validators/instance_changes.ts:23]: WDYT should be the new message?

File: packages/netsuite-adapter/src/adapter.ts [hunk 5]
```diff
@@ -270,14 +272,22 @@ export default class NetsuiteAdapter implements AdapterOperations {
...
-      const type = isCustomTypeName(customizationInfo.typeName)
-        ? customTypes[customizationInfo.typeName].type
-        : additionalTypes[customizationInfo.typeName]
-      return type
-        ? createInstanceElement(customizationInfo, type, this.getElemIdFunc, serverTime)
-        : undefined
-    }).filter(isInstanceElement).toArray()
+    const [customRecordTypeInstances, instances] = _.partition(
+      await awu(customizationInfos).map(customizationInfo => {
+        const type = isCustomTypeName(customizationInfo.typeName)
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/adapter.ts:276]: maybe `isStandardTypeName` (and all other places using `customType` somehow) ?

[the remark this answers is not in the document]
Comment (omrilit) [reply packages/netsuite-adapter/src/adapter.ts:276]: Sounds great!

File: packages/netsuite-adapter/src/change_validators/safe_deploy.ts [hunk 6]
```diff
@@ -70,64 +68,82 @@ const getIdentifingValuesByType = async (
...
+}
 
-const getMatchingServiceInstances = async (
-  baseInstances: InstanceElement[],
+const getElementToCompare = (element: Element): Element => {
+  const clonedElement = element.clone()
+  if (isInstanceElement(clonedElement)) {
+    delete clonedElement.value[LAST_FETCH_TIME]
+  }
+  delete clonedElement.annotations[LAST_FETCH_TIME]
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/change_validators/safe_deploy.ts]: changed.

File: packages/netsuite-adapter/src/custom_records/custom_record_type.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        annotations: { [CORE_ANNOTATIONS.HIDDEN_VALUE]: true },
+      },
+    },
+    annotationRefsOrTypes: {
+      ...annotationRefsOrTypes,
+      source: BuiltinTypes.HIDDEN_STRING,
+    },
+    annotations: {
+      ...instance.value,
+      [SOURCE]: 'soap',
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/custom_records/custom_record_type.ts:64]: The custom record instances will be fetched using SOAP.

File: packages/netsuite-adapter/src/custom_records/custom_record_type.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      [METADATA_TYPE]: CUSTOM_RECORD_TYPE,
+    },
+    path: [NETSUITE, CUSTOM_RECORDS_PATH, instance.value[SCRIPT_ID]],
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/custom_records/custom_record_type.ts:67]: yes

File: packages/netsuite-adapter/src/client/client.ts [hunk 9]
```diff
@@ -148,36 +149,65 @@ export default class NetsuiteClient {
...
-    const changedInstances = changes.map(getChangeData)
-    const failedElemIDs = new Set(changedInstances
-      .filter(inst => error.failedObjects.has(inst.value[SCRIPT_ID]))
-      .map(inst => inst.elemID.getFullName()))
+    const changeData = changes.map(getChangeData)
+    const failedElemIDs = new Set(changeData
+      .filter(elem => (
+        isInstanceElement(elem) && error.failedObjects.has(elem.value[SCRIPT_ID])
+      ) || (
+        isField(elem) && error.failedObjects.has(elem.parent.annotations[SCRIPT_ID])
```
[the remark this answers is not in the document]
Comment (dantal4) [reply packages/netsuite-adapter/src/client/client.ts:159]: I couldn't find more usages of this code. Can you point it out?

[the remark this answers is not in the document]
Comment (omrilit) [reply packages/netsuite-adapter/src/client/client.ts:159]: Maybe I was wrong, if you couldn't find it then ignore it

File: packages/netsuite-adapter/src/group_changes.ts [hunk 10]
```diff
@@ -49,19 +47,24 @@ export const SUITEAPP_FILE_CABINET_GROUPS = [
...
 }
 
 const getChangeGroupIdsWithoutSuiteApp: ChangeGroupIdFunction = async changes => {
   const isSdfChange = (change: Change): boolean => {
     const changeData = getChangeData(change)
-    return isInstanceElement(changeData)
-      && (isCustomTypeName(changeData.elemID.typeName)
-        || fileCabinetTypesNames.has(changeData.elemID.typeName)
-        || isSDFConfigTypeName(changeData.elemID.typeName))
+    return isStandardOrCustomRecordType(changeData)
```
Comment (omrilit) [inline packages/netsuite-adapter/src/group_changes.ts:65]: Do you think it should be renamed to isStandardInstanceOrCustomRecordType?

**Review by omrilit — APPROVED** [review]
Approving to save cycle.
Do we need to have the INDEX defined in the type of the fields?
It would require creating `netsuite,number` type instead of `number` field for that.
If we decide not to add it, maybe we should have a change validator to warn about it?

**Pull request merged**